src/devtools/useAtomsDevtools.ts

```diff
@@ -0,0 +1,212 @@
+
+  if (!store[DEV_SUBSCRIBE_STATE]) {
+    throw new Error('useAtomsSnapshot can only be used in dev mode.')
+  }
+
+  const isTimeTraveling = useRef(false)
+  const isRecording = useRef(true)
+  const devtools = useRef<ConnectionResult & { shouldInit?: boolean }>()
+
+  const snapshots = useRef<AtomsSnapshot[]>([])
+
+  useEffect(() => {
+    if (extension) {
+      const connection = extension.connect({ name })
+      const devtoolsUnsubscribe = connection.subscribe((message: Message) => {
+        switch (message.type) {
+          case 'DISPATCH':
+            switch (message.payload?.type) {
+              case 'RESET':
+                // Todo
+                return
+              case 'COMMIT': {
+                const lastSnapshot = snapshots.current[
+                  snapshots.current.length - 1
+                ] as AtomsSnapshot
+
+                const serializedSnapshot = serializeSnapshot(lastSnapshot)
+
+                connection.init({
+                  values: serializedSnapshot,
+                  dependencies: getDependencies(store, lastSnapshot),
+                })
+                snapshots.current.length = 0
+                return
+              }
+
+              case 'JUMP_TO_ACTION': {
+                isTimeTraveling.current = true
+
+                const currentSnapshot = snapshots.current[
+                  message.payload.actionId - 1
+                ] as AtomsSnapshot
+
+                goToSnapshot(currentSnapshot)
+                return
+              }
+
+              case 'PAUSE_RECORDING':
+                isRecording.current = !isRecording.current
+                return
+            }
+        }
+      })
+
+      devtools.current = connection
+      devtools.current.shouldInit = true
+      return devtoolsUnsubscribe
+    }
+  }, [store, extension, goToSnapshot, name])
+
+  useEffect(() => {
+    if (!devtools.current) {
+      return
+    }
+    if (devtools.current.shouldInit) {
+      devtools.current.init(undefined)
+      devtools.current.shouldInit = false
+      return
+    }
+
+    if (!snapshot.size) {
+      return
+    }
+
+    if (isTimeTraveling.current) {
+      queueMicrotask(() => {
+        isTimeTraveling.current = false
+      })
+    } else if (isRecording.current) {
+      queueMicrotask(() => {
+        batchedLog(devtools.current!, snapshots.current!, snapshot, store)
+      })
+    }
+  }, [snapshot, store])
+}
+
+const batchesQueue: {
+  values: any
+  dependencies: any
+}[] = []
+const snapshotsQueue: AtomsSnapshot[] = []
+
+const batchedLog = (
+  devtools: ConnectionResult,
+  snapshots: AtomsSnapshot[],
+  snapshot: AtomsSnapshot,
+  store: Store
+) => {
+  const serializedSnapshot = serializeSnapshot(snapshot)
+
+  const state = {
+    values: serializedSnapshot,
+    dependencies: getDependencies(store, snapshot),
+  }
+
+  batchesQueue.push(state)
+  snapshotsQueue.push(snapshot)
+  const batchesLength = batchesQueue.length
+  setTimeout(() => {
```

Comment:
This isn't what I expected. Why I wonder why `useEffect` doesn't batch... I might be missing something.

Reply:
which part you mean it doesn't batch?